### DISCUSSION TITLE
Preview fix that you implamented.

### DISCUSSION THREAD

**Discussion by askusbloodfist** [Vocaluxe/Vocaluxe#621] [General]
Hi i have a question. what did you guys do in oder to get the Previews working always?
Ive notice in Ultrastar general softwere this is not working on all cases. So you guys must have done something that made this work?

**Comment by marwin89** [comment]
@Stefan1200-de do you know something about preview functionality in vocaluxe? I read in different ultrastar forums about preview issues too. only vocaluxe seems to handle this?

**Comment by Stefan1200-de** [comment]
> @Stefan1200-de do you know something about preview functionality in vocaluxe? I read in different ultrastar forums about preview issues too. only vocaluxe seems to handle this?

You talk about the song preview on the song selection screen?
Well, last year I looked into that code, because I had problems too (the preview of some songs starts at pos 0), but I never understand it completely. Since then I added #previewstart in every song in my collection. :)

**Comment by flokuep** [comment]
It's not easy to answer that questions. I guess the reason for not working preview could lay in the implementation itself, the underlying sound library or even the audio file that should be played. I cannot say what Vocaluxe does in difference to the other games.

In Vocaluxe we have three (I guess) different sound libraries that could be used for playback: GStreamer, PortAudio and OpenAL. Implementation differs from what the libraries are providing and needing for skipping some time within an audio file.

Another part of the preview stuff is to distinguish which position of the song should be used for the preview if there is none set in the text file explicitly.

Vocaluxe uses the medley start beat, as for medleys the refrain is often used and that's also a good song part for the preview. The medley start beat could come again from a tag within the textfile. If there is none, Vocaluxe tries to find a good medley piece by looking at the lyrics and their frequency (more details in CSong._CalcMedley). If this isn't possible, it fallbacks to the song start.

@Stefan1200-de Either there is a problem with the playback lib you're using or there is a bug withing that medley-calc logic (I wouldn't be suprised, there is some room for improvements within that code).
